### PR TITLE
new(Lightbox): Add Lightbox component

### PR DESCRIPTION
**Pull request by piewell**
to: @milesj @stefhatcher @hayes @alecklandgraf
cc: @Alphy11 

## Description

Add a Lightbox component to Lunar, with optional zoom and rotate features.

## Motivation and Context

A Lightbox component is needed for multiple teams. One team requires a zoom and rotate feature, which can be added by using the props `showZoomControls` and `showRotateControls`. The prop `images` takes an object with an `src`, `alt`, and optional `aside` value. 

```
// Example image prop value:
const mockImages = [
  { src: moon, alt: 'moon' },
  { src: space, alt: 'space' },
  { src: stars, alt: 'stars' },
];

// With sidebar content:
const mockImagesWithAside = [
  { src: moon, alt: 'moon', aside: <AsideDemo title="Moon" /> },
  { src: space, alt: 'space', aside: <AsideDemo title="Space" /> },
  { src: stars, alt: 'stars', aside: <AsideDemo title="Stars" /> },
];
```

All examples are in Storybook.

## Testing

[x] unit tests
[x] storybook

## Screenshots

Base Lightbox:
Screen Shot 2019-09-12 at 4 50 21 PM

With optional sidebar:
Screen Shot 2019-09-12 at 4 50 48 PM

With zoom, rotate, and sidebar:
Screen Shot 2019-09-18 at 10 47 31 AM

## Checklist

- My code follows the style guide of this project.
- I have updated or added documentation accordingly.
- I have read the CONTRIBUTING document.

### CODE REVIEW

File: packages/core/src/components/Lightbox/LightboxHeader.tsx [hunk 1]
```diff
@@ -0,0 +1,163 @@
+import React from 'react';
```
Comment (milesj) [inline packages/core/src/components/Lightbox/LightboxHeader.tsx]: File can just be called `Header.tsx`

File: packages/core/src/components/Lightbox/LightboxHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+import Button from '../Button';
+import Pagination from '../Pagination';
+import T from '../Translate';
+import withStyles, { WithStylesProps } from '../../composers/withStyles';
+import { ARROW_RIGHT, ARROW_LEFT } from '../../keys';
+import { ZoomControls, RotateControls } from '../ImageViewer';
+import Spacing from '../Spacing';
+
+export type LightboxHeaderProps = {
```
Comment (milesj) [inline packages/core/src/components/Lightbox/LightboxHeader.tsx]: `Props`. We don't use the full name in Lunar.

File: packages/core/src/components/Lightbox/LightboxImage.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
+import React from 'react';
```
Comment (milesj) [inline packages/core/src/components/Lightbox/LightboxImage.tsx]: `Image.tsx`

File: packages/core/src/components/Lightbox/LightboxImage.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import withStyles, { WithStylesProps } from '../../composers/withStyles';
+import ResponsiveImage from '../ResponsiveImage';
+import Loader from '../Loader';
+import ImageViewer from '../ImageViewer';
+
+export type LightboxImageProps = {
```
Comment (milesj) [inline packages/core/src/components/Lightbox/LightboxImage.tsx]: `Props`

File: packages/core/src/components/Lightbox/index.tsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
+import React from 'react';
+import Sheet from '../Sheet';
+import LightboxImage, { LightboxImageProps } from './LightboxImage';
+import LightboxHeader from './LightboxHeader';
+
+export type LightboxProps = {
```
Comment (milesj) [inline packages/core/src/components/Lightbox/index.tsx]: `Props`

File: packages/core/src/components/Lightbox/index.tsx [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  startIndex?: number;
+  /** Show zoom controls. */
+  showZoomControls?: boolean;
+  /** Show rotate controls. */
+  showRotateControls?: boolean;
+  /** Function to run on close. */
+  onClose: () => void;
+};
+
+export type LightboxState = {
```
Comment (milesj) [inline packages/core/src/components/Lightbox/index.tsx]: `State`

File: packages/core/src/components/Lightbox/index.tsx [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  };
+
+  state = {
+    activeIndex: 0,
+    hideAside: false,
+    scale: 1,
+    rotation: 0,
+  };
+
+  preloadedUrls: Map<string, boolean | HTMLImageElement>;
```
Comment (milesj) [inline packages/core/src/components/Lightbox/index.tsx]: `preloadedUrls = new Map<string, boolean | HTMLImageElement>();` and then you can remove the constructor.

File: packages/core/src/components/Lightbox/index.tsx [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  componentDidUpdate() {
+    if (this.props.images.length > 1) {
+      // prefetch the next image when `images` or `activeIndex` changes
+      this.preloadImages();
+    }
+  }
+
+  componentWillUnmount() {
+    // We no longer need these images, so let's remove them so they can be garbage collected.
+    delete this.preloadedUrls;
```
Comment (milesj) [inline packages/core/src/components/Lightbox/index.tsx]: `clear()`?

File: packages/core/src/components/Lightbox/index.tsx [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  preloadImages() {
+    const { images } = this.props;
+    const { activeIndex } = this.state;
+    const totalImages = images.length;
+    const startIndex = activeIndex + 1; // start direction is going forward (+1)
+    const endIndex = activeIndex + 2;
+    const indices = Array.from({ length: endIndex - startIndex }, (_, i) => i + startIndex);
```
Comment (milesj) [inline packages/core/src/components/Lightbox/index.tsx]: `const indices = [startIndex, endIndex]`? Confused what the from() is actually for.

File: packages/core/src/components/Lightbox/LightboxImage.tsx (hunk 3 above)
Comment (piewell) [reply packages/core/src/components/Lightbox/LightboxImage.tsx]: 👍 I'll need to use the name `LightboxImage` in the Lightbox itself to differentiate from the plain javascript `Image` component

Comment (milesj) [reply packages/core/src/components/Lightbox/LightboxImage.tsx]: You can rename on import: `import LightboxImage from './Image';`. The file itself can be `Image.tsx`.

**Comment by piewell** [comment]
All checks passed! PTAL @stefhatcher @milesj

**Comment by milesj** [comment]
@piewell Can you rebase and update your story to the new format?

**Comment by piewell** [comment]
@milesj Fixed! Happo needs to be approved, PTAL

**Pull request merged**